runtime/src/accounts_db.rs

```diff
@@ -2109,28 +2110,33 @@ impl AccountsDb {
             let do_clean_scan = || {
                 pubkeys
                     .par_chunks(4096)
-                    .map(|pubkeys: &[Pubkey]| {
+                    .map(|pubkeys: &[(Pubkey, bool)]| {
                         let mut purges_zero_lamports = HashMap::new();
                         let mut purges_old_accounts = Vec::new();
                         let mut found_not_zero = 0;
                         let mut not_found_on_fork = 0;
                         let mut missing = 0;
                         let mut useful = 0;
                         self.accounts_index.scan(
-                            pubkeys,
+                            pubkeys.iter(),
                             max_clean_root,
                             // return true if we want this item to remain in the cache
-                            |exists, slot_list, index_in_slot_list, pubkey, ref_count| {
+                            |exists,
+                             slot_list,
+                             index_in_slot_list,
+                             pubkey,
+                             ref_count,
+                             zero_lamport| {
                                 let mut useless = true;
                                 if !exists {
                                     missing += 1;
                                 } else {
                                     match index_in_slot_list {
                                         Some(index_in_slot_list) => {
                                             // found info relative to max_clean_root
-                                            let (slot, account_info) =
+                                            let (slot, _account_info) =
                                                 &slot_list[index_in_slot_list];
-                                            if account_info.lamports == 0 {
+                                            if zero_lamport {
```

Comment:
Note, this seems to be the crux of the change

Reply:
this has been totally reworked